Assets/MixedRealityToolkit/Interfaces/InputSystem/IMixedRealityRaycastProvider.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Physics;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// Interface to handle raycasts into the scene. Used by FocusProvider to perform ray and sphere cast queries for pointers.
+    /// </summary>
+    /// <remarks>
+    /// Implementations of IMixedRealityRaycastProvider would likely use Unity's physics system to get hit results from Colliders. However,
+    /// in a custom implementation, the raycast does not have to rely only on Unity-based Colliders to provide hit results, e.g. a
+    /// GameObject may use a different mechanism for raycasting, and with a custom implementation, it could be included in the hit result.
+    /// </remarks>
+    public interface IMixedRealityRaycastProvider : IMixedRealityDataProvider
+    {
+        /// <summary>
+        /// Performs a raycast using the specified <see cref="Microsoft.MixedReality.Toolkit.Physics.RayStep"/>.
+        /// </summary>
+        /// <returns>Whether or not the raycast hit something.</returns>
+        bool Raycast(RayStep step, LayerMask[] prioritizedLayerMasks, out MixedRealityRaycastHit hitInfo);
```

Comment:
One more level of configuration that can be helpful is being able to configure our own struct for capturing the RaycastHitResult. Some of our apps also captured the hit type - Hologram, SR, EmptySpace. I am not sure if it makes sense to capture this information in the default MixedRealityRaycastHit struct or just have the option of an interface so that developers creating their custom raycastproviders also have the option of custom raycasthit results.

Reply:
Since the MixedRealityRaycastHit is a class, it could be extended to provide additional hit data. In a custom FocusProvider implementation, you could then cast the hit back to the type you expect and get the extra data off of it.